dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx

```diff
@@ -139,32 +139,77 @@ cpdef utils.dpnp_descriptor dpnp_diag(utils.dpnp_descriptor v, int k):
     return result
 
 
-cpdef utils.dpnp_descriptor dpnp_geomspace(start, stop, num, endpoint, dtype, axis):
-    cdef shape_type_c obj_shape = utils._object_to_tuple(num)
-    cdef utils.dpnp_descriptor result = utils_py.create_output_descriptor_py(obj_shape, dtype, None)
+def _copy_by_usm_type(x, dtype, usm_type, sycl_queue):
+    if not hasattr(x, "usm_type"):
+        result = dpnp.asarray(x, dtype=dtype, usm_type=usm_type, sycl_queue=sycl_queue)
+    elif usm_type == "host" and x.usm_type != "host" or usm_type == "shared" and x.usm_type == "device":
+        result = dpnp.asarray(x, dtype=dtype, usm_type=usm_type, sycl_queue=sycl_queue)
+    else:
+        result = dpnp.asarray(x, dtype=dtype, sycl_queue=sycl_queue)
+    return result
+
+
+def _get_start_stop(start, stop, device=None, usm_type=None, sycl_queue=None):
+    usm_type_alloc, sycl_queue_alloc = utils_py.get_usm_allocations([start, stop])
+
+    # Get sycl_queue.
+    if sycl_queue is None and device is None:
+        sycl_queue = sycl_queue_alloc
+    sycl_queue_normalized = dpnp.get_normalized_queue_device(sycl_queue=sycl_queue, device=device)
+
+    # Get usm_type.
+    if usm_type is None:
+        usm_type = "device" if usm_type_alloc is None else usm_type_alloc
+
+    # Get dtype.
+    if not hasattr(start, "dtype") and not dpnp.isscalar(start):
+        start = dpnp.asarray(start, usm_type=usm_type, sycl_queue=sycl_queue_normalized)
+    if not hasattr(stop, "dtype") and not dpnp.isscalar(stop):
+        stop = dpnp.asarray(stop, usm_type=usm_type, sycl_queue=sycl_queue_normalized)
+    dt = numpy.result_type(start, stop, float(0))
+    dt = utils_py.map_dtype_to_device(dt, sycl_queue_normalized.sycl_device)
+
+    # Copy arrays if needed with current dtype, usm_type and sycl_queue.
+    # Do not need to copy usm_ndarray by usm_type if it is not explicitly stated.
+    _start = _copy_by_usm_type(start, dtype=dt, usm_type=usm_type, sycl_queue=sycl_queue_normalized)
+    _stop = _copy_by_usm_type(stop, dtype=dt, usm_type=usm_type, sycl_queue=sycl_queue_normalized)
+    return _start, _stop
+
+
+def dpnp_geomspace(start, stop, num, dtype=None, device=None, usm_type=None, sycl_queue=None, endpoint=True, axis=0):
+    num = operator.index(num)
+    if num < 0:
+        raise ValueError("Number of points must be non-negative")
+
+    _start, _stop = _get_start_stop(start, stop, device, usm_type, sycl_queue)
+
+    if dtype is None:
+        dtype = _start.dtype
 
     if endpoint:
         steps_count = num - 1
     else:
         steps_count = num
-
-    # if there are steps, then fill values
     if steps_count > 0:
-        step = dpnp.power(dpnp.float64(stop) / start, 1.0 / steps_count)
-        mult = step
-        for i in range(1, result.size):
-            result.get_pyobj()[i] = start * mult
-            mult = mult * step
+        step = (_stop / _start) ** (1.0 / steps_count)
     else:
         step = dpnp.nan
```

Comment:
I'm not sure if we need to consider the use case, but dpnp generates different array when
```
# yn = numpy.geomspace(-1, 1, num=3, dtype=complex)
#: y = dpnp.geomspace(-1, 1, num=3, dtype=complex)
# assert_allclose(yn, y.asnumpy())
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[56], line 1
----> 1 assert_allclose(yn, y.asnumpy())

    [... skipping hidden 2 frame]

File /localdisk/work/antonvol/soft/miniconda3/envs/dpnp_py39_ext/lib/python3.9/site-packages/numpy/testing/_private/utils.py:745, in assert_array_compare.<locals>.func_assert_same_pos(x, y, func, hasval)
    740 if bool_(x_id == y_id).all() != True:
    741     msg = build_err_msg([x, y],
    742                         err_msg + '\nx and y %s location mismatch:'
    743                         % (hasval), verbose=verbose, header=header,
    744                         names=('x', 'y'), precision=precision)
--> 745     raise AssertionError(msg)
    746 # If there is a scalar, then here we know the array has the same
    747 # flag as it everywhere, so we should return the scalar flag.
    748 if isinstance(x_id, bool) or x_id.ndim == 0:

AssertionError:
Not equal to tolerance rtol=1e-07, atol=0

x and y nan location mismatch:
 x: array([-1.000000e+00+0.j,  6.123234e-17+1.j,  1.000000e+00+0.j])
 y: array([-1.+0.j, nan+0.j,  1.+0.j])

```